GitCommands/Git/GitModule.cs

```diff
@@ -3523,6 +3523,60 @@ public MemoryStream GetFileStream(string blob)
             // TODO why return a stream here? should just return a byte[]
 
             try
+            {
+                MemoryStream memoryStream = CatFile();
+                if (!IsTrackedByLFS(memoryStream))
+                {
+                    return memoryStream;
+                }
+                else
+                {
+                    return LFSSmudge(memoryStream);
+                }
+            }
+            catch (Win32Exception ex)
+            {
+                Trace.WriteLine(ex);
+                return null;
+            }
+
+            MemoryStream LFSSmudge(MemoryStream cleanStream)
+            {
+                var args = new GitArgumentBuilder("lfs") { "smudge" };
+                using (var process = RunGitCmdDetached(args, redirectInput: true, redirectOutput: true))
+                {
+                    cleanStream.CopyTo(process.StandardInput.BaseStream);
+                    process.StandardInput.Close();
+                    cleanStream.Dispose();
+                    var dirtyStream = new MemoryStream();
+                    process.StandardOutput.BaseStream.CopyTo(dirtyStream);
+                    process.WaitForExit();
+                    dirtyStream.Position = 0;
+                    return dirtyStream;
+                }
+            }
+
+            // This is a heuristic because there's no way to get this information *exactly* without checking
+            // .gitattributes from the relevant commit, and that seems painful
```

Comment:
Exactly what you should do.  LFS may change the header.  The correct way is to use .gitattributes.  https://github.com/git-lfs/git-lfs/issues/2748#issuecomment-347938047

Reply:
Oh, this is useful! But still, there is an issue that at the point of `GetFileStream` I only have the blob, not the path of the file, nor the sha of the tree. Do you have any suggestions on how I could get this information here?

Reply:
Is it safe to return the `process.StandardOutput` stream? Is it possible it will never get consumed and the LFS process is left hanging?  
`GetFileStream` was returning a `MemoryStream` before I touched it, I didn't want to change that because I'm not very familiar with the code yet.

Reply:
Good point.  It was just something I noticed.  Don't worry about it now.

Reply:
> Is it safe to return the process.StandardOutput stream?
No.  It should be copied.  Don't worry about the memorystream stuff.  That is a different issue. 
 
There are only two references.  
![image](https://user-images.githubusercontent.com/3629489/51982907-42a64300-2465-11e9-86b1-5a1e0e58034e.png)

I'm trying to think on how to get the path from the blob.  We need at least the commit sha and the blob id.  I'll get back to you once I have a solution.

Reply:
We can shorten this to something real simple. 

```bash
git cat-file --filters <blob Sha>
```

https://git-scm.com/docs/git-cat-file#git-cat-file---filters

Reply:
Dang it.  **current working tree**

Reply:
The more I think about it, the more it seems to me that my original issue was just a symptom of the fact that files are *in general* not filtered when accessed through the tree. Should they be? [here](https://github.com/gitextensions/gitextensions/issues/6146#issuecomment-456462655) it seems that `git difftool` thinks they should...  
Does `git difftool` filter the files correctly anyway? Does it perhaps use the current working tree filter rules?

This gets invoked when diff the .png:  
`difftool --gui --no-prompt -M -C "<sha_rev_1>" "<sha_rev_2>" -- "<path_to_file>"`

Reply:
Yes. Difftool needs the actual contents of the file so filters are run. Read up on git filters.

Reply:
Yes, I get that.

What I'm wondering is: *should* there be a similar process of applying filters *every* time GitExtensions shows a file (`cat-file` or whatever).  

Note that, as @gerhardol said, filters are not applied on git-diff.

On another note, I see it would be possible to pass all the required information down to `GetFileStream` in every case except when it is invoked from `FormVerify.saveAsToolStripMenuItem_Click`.

Reply:
We can deal with when it should be run after it works. I think there are
reasons why diff doesn't run filters. If you think about it. Diff should
just show the pointer.  You also don't want to download every single or
just about lfs file just because you ran diff. Difftool makes sense because
the user expects the actual contents. Don't want to be editing a pointer
file

Reply:
--textconv
--no-textconv
Allow (or disallow) external text conversion filters to be run when comparing binary files. See gitattributes[5] for details. Because textconv filters are typically a one-way conversion, the resulting diff is suitable for human consumption, but cannot be applied. For this reason, textconv filters are enabled by default only for git-diff[1] and git-log[1], but not for git-format-patch[1] or diff plumbing commands.

Reply:
--textconv
--no-textconv
Allow (or disallow) external text conversion filters to be run when comparing binary files. See gitattributes[5] for details. Because textconv filters are typically a one-way conversion, the resulting diff is suitable for human consumption, but cannot be applied. For this reason, textconv filters are enabled by default only for git-diff[1] and git-log[1], but not for git-format-patch[1] or diff plumbing commands.

Reply:
OK.  
So, what's my next step? Do I try to get the commit sha and file path down to `GetFileStream` or do you want me to do something else? Or do you want me to do nothing for now?

Reply:
I know the lfs team has to get their work done and it will have to be pulled into git itself. Also, we will have to do a version check on a version that supports the official --check argument. Good that we got the ball rolling on that.

I hate that we have to read the header but it looks like it that is the best bet at the moment.  We will have to maintain this and handle version changes of the header. I would look at how git lfs reads the pointer and try to match as close as possible to that. 

Note for later  

Only do this if version doesn't support --check option when it is available.

Reply:
I know the lfs team has to get their work done and it will have to be pulled into git itself. Also, we will have to do a version check on a version that supports the official --check argument. Good that we got the ball rolling on that.

I hate that we have to read the header but it looks like it that is the best bet at the moment.  We will have to maintain this and handle version changes of the header. I would look at how git lfs reads the pointer and try to match as close as possible to that. 

Note for later  

Only do this if version doesn't support --check option when it is available.

Reply:
> Note that, as @gerhardol said, filters are not applied on git-diff.

It could be possible to force the filters to run also for git-diff, if this is a file tracked by LFS, similar to what is planned for Save As.

> git cat-file --filters ...
> Dang it. current working tree

Is this not good enough?
The main argument against is if a file has been added to LFS at a later stage so maybe no.

The program is named GitExtensions for a reason.

Reply:
No. The issue is working directory . gitattributes could not have that filter. We want lfs to say yes you have a pointer file. We have pushed that idea at the lfs team. I can see holes in the logic and results. I want to be able to say this is a pointer without any gitattributes issues.

Reply:
--filters  

Show the content as converted by the filters configured in the current working tree for the given <path> (i.e. smudge filters, end-of-line conversion, etc). In this case, <object> has to be of the form <tree-ish>:<path>, or :<path>.

Reply:
Ok. Game plan is this. 

Use a version of the command mentioned here. https://github.com/git-lfs/git-lfs/pull/3501#issuecomment-459021407

We will have to wait till lfs supports it. For now use the header like you are.